File: zotprep/search/pubmed.py
"""PubMed provider — deterministic first, free-text second.

The important piece is `ecitmatch`, NCBI's citation matcher. Given
journal|year|volume|first_page|author it returns the exact PMID or NOT_FOUND.
Because Vancouver references already carry NLM journal abbreviations, volume and
first page, this turns most biomedical references into a *lookup* rather than a
search — no relevance ranking, no ambiguity, nothing to score.

It also works with an empty author field, which is what rescues the consortium
references (GBD / India State-Level Collaborators) whose "first author" is not a
personal surname at all.
"""
from __future__ import annotations

import re

import httpx

from ..models import Candidate, ParsedRef
from ..utils import surname_of
from .base import get

EUTILS = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils"
TOOL = "zotprep"


def _to_candidate(rec: dict) -> Candidate:
    authors, corporate = [], None
    for a in rec.get("authors") or []:
        nm = a.get("name") or ""
        if not nm:
            continue
        if a.get("authtype") == "CollectiveName" or re.search(
            r"\b(collaborators?|group|consortium|committee|network)\b", nm, re.I
        ):
            corporate = corporate or nm
            continue
        authors.append(surname_of(nm))
    doi = next(
        (i["value"] for i in rec.get("articleids") or [] if i.get("idtype") == "doi"), None
    )
    ym = re.search(r"\b(19|20)\d{2}\b", rec.get("pubdate") or "")
    pages = rec.get("pages") or ""
    fp, _, lp = pages.partition("-")
    return Candidate(
        provider="pubmed",
        title=(rec.get("title") or "").rstrip("."),
        doi=doi,
        pmid=str(rec.get("uid") or "") or None,
        authors=authors,
        corporate=corporate,
        year=int(ym.group()) if ym else None,
        journal=rec.get("fulljournalname") or rec.get("source") or "",
        journal_abbrev=rec.get("source") or "",
        volume=rec.get("volume") or None,
        issue=rec.get("issue") or None,
        first_page=fp.strip() or None,
        last_page=lp.strip() or None,
        raw=rec,
    )


async def _summaries(client: httpx.AsyncClient, pmids: list[str], email: str) -> list[Candidate]:
    if not pmids:
        return []
    data = await get(
        client,
        "pubmed",
        f"{EUTILS}/esummary.fcgi",
        params={
            "db": "pubmed",
            "id": ",".join(pmids),
            "retmode": "json",
            "tool": TOOL,
            "email": email,
        },
    )
    result = (data or {}).get("result") or {}
    return [_to_candidate(result[p]) for p in pmids if isinstance(result.get(p), dict)]


async def citation_match(client: httpx.AsyncClient, ref: ParsedRef, email: str) -> str | None:
    """Deterministic PMID lookup from the reference's own locator fields.

    Tries with the first author and again with the author field blank, because
    the author spelling is the least reliable part of the tuple (initial order,
    accents, consortium names) while journal/year/volume/page are exact.
    """
    if not (ref.journal and ref.year and ref.volume and ref.first_page):
        return None
    author_variants = [""]
    if ref.lead_author:
        author_variants.insert(0, ref.lead_author)
    for author in author_variants:
        bdata = "|".join(
            [ref.journal, str(ref.year), ref.volume, ref.first_page, author, f"z{ref.n}"]
        ) + "|"
        text = await get(
            client,
            "pubmed",
            f"{EUTILS}/ecitmatch.cgi",
            params={"db": "pubmed", "retmode": "xml", "bdata": bdata},
            expect_json=False,
        )
        if not text:
            continue
        tail = text.strip().split("|")[-1].strip()
        if tail.isdigit():
            return tail
    return None


async def search(client: httpx.AsyncClient, ref: ParsedRef, email: str) -> list[Candidate]:
    """ecitmatch first; fall back to a title-field esearch."""
    pmids: list[str] = []
    exact = await citation_match(client, ref, email)
    if exact:
        pmids.append(exact)

    if ref.title:
        # [Title] restricts to the title field, which is far more precise than
        # the old code's untagged whole-reference term.
        term = f'"{ref.title}"[Title]'
        data = await get(
            client,
            "pubmed",
            f"{EUTILS}/esearch.fcgi",
            params={
                "db": "pubmed",
                "term": term,
                "retmax": 8,
                "retmode": "json",
                "tool": TOOL,
                "email": email,
            },
        )
        found = ((data or {}).get("esearchresult") or {}).get("idlist") or []
        if not found:
            # loosen to all-fields when the exact title phrase isn't indexed
            data = await get(
                client,
                "pubmed",
                f"{EUTILS}/esearch.fcgi",
                params={
                    "db": "pubmed",
                    "term": f"{ref.title} {ref.lead_author or ''}".strip(),
                    "retmax": 8,
                    "retmode": "json",
                    "tool": TOOL,
                    "email": email,
                },
            )
            found = ((data or {}).get("esearchresult") or {}).get("idlist") or []
        pmids += [p for p in found if p not in pmids]

    cands = await _summaries(client, pmids[:10], email)
    for c in cands:
        if exact and c.pmid == exact:
            # flag for the scorer: this one came from the deterministic matcher
            c.provider = "pubmed:ecitmatch"
            c.providers = {"pubmed:ecitmatch", "pubmed"}
    return cands


async def by_pmid(client: httpx.AsyncClient, pmid: str, email: str) -> Candidate | None:
    cands = await _summaries(client, [pmid], email)
    return cands[0] if cands else None

File: web/src/search/pubmed.js
/**
 * Port of zotprep/search/pubmed.py — deterministic first, free-text second.
 *
 * The important piece is `ecitmatch`, NCBI's citation matcher. Given
 * journal|year|volume|first_page|author it returns the exact PMID or NOT_FOUND.
 * Because Vancouver references already carry NLM journal abbreviations, volume
 * and first page, this turns most biomedical references into a *lookup* rather
 * than a search — no relevance ranking, no ambiguity, nothing to score.
 *
 * It also works with an empty author field, which is what rescues the consortium
 * references (GBD / India State-Level Collaborators) whose "first author" is not
 * a personal surname at all.
 */
import { Candidate } from '../models.js';
import { surnameOf } from '../utils.js';
import { get } from './base.js';
import { partition } from './crossref.js';

export const EUTILS = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils';
// NCBI's `tool` parameter, which identifies the client in their logs. The CLI
// sends "zotprep"; this is a different application talking to the same API, so
// it says so. Purely identification — it does not affect any result.
const TOOL = 'z-link';

const COLLECTIVE_RE = /(?<![\p{L}\p{N}_])(collaborators?|group|consortium|committee|network)(?![\p{L}\p{N}_])/iu;
const YEAR_RE = /(?<![\p{L}\p{N}_])(19|20)\d{2}(?![\p{L}\p{N}_])/u;

function toCandidate(rec) {
  const authors = [];
  let corporate = null;
  for (const a of (rec.authors || [])) {
    const nm = a.name || '';
    if (!nm) continue;
    if (a.authtype === 'CollectiveName' || COLLECTIVE_RE.test(nm)) {
      corporate = corporate || nm;
      continue;
    }
    authors.push(surnameOf(nm));
  }
  const doiEntry = (rec.articleids || []).find((i) => i.idtype === 'doi');
  const ym = YEAR_RE.exec(rec.pubdate || '');
  const [fp, , lp] = partition(rec.pages || '', '-');
  return new Candidate({
    provider: 'pubmed',
    title: (rec.title || '').replace(/\.+$/, ''),
    doi: doiEntry ? doiEntry.value : null,
    pmid: String(rec.uid || '') || null,
    authors,
    corporate,
    year: ym ? parseInt(ym[0], 10) : null,
    journal: rec.fulljournalname || rec.source || '',
    journal_abbrev: rec.source || '',
    volume: rec.volume || null,
    issue: rec.issue || null,
    first_page: fp.trim() || null,
    last_page: lp.trim() || null,
    raw: rec,
  });
}

async function summaries(pmids, email, opts = {}) {
  if (!pmids.length) return [];
  const data = await get('pubmed', `${EUTILS}/esummary.fcgi`, {
    params: {
      db: 'pubmed', id: pmids.join(','), retmode: 'json', tool: TOOL, email,
    },
    ...opts,
  });
  const result = (data || {}).result || {};
  return pmids
    .filter((p) => result[p] && typeof result[p] === 'object')
    .map((p) => toCandidate(result[p]));
}

/**
 * Deterministic PMID lookup from the reference's own locator fields.
 *
 * Tries with the first author and again with the author field blank, because the
 * author spelling is the least reliable part of the tuple (initial order,
 * accents, consortium names) while journal/year/volume/page are exact.
 */
export async function citationMatch(ref, _email, opts = {}) {
  if (!(ref.journal && ref.year && ref.volume && ref.first_page)) return null;
  const authorVariants = [''];
  if (ref.lead_author) authorVariants.unshift(ref.lead_author);
  for (const author of authorVariants) {
    const bdata = [ref.journal, String(ref.year), ref.volume, ref.first_page,
      author, `z${ref.n}`].join('|') + '|';
    const text = await get('pubmed', `${EUTILS}/ecitmatch.cgi`, {
      params: { db: 'pubmed', retmode: 'xml', bdata },
      expectJson: false,
      ...opts,
    });
    if (!text) continue;
    const tail = text.trim().split('|').pop().trim();
    if (/^\d+$/.test(tail)) return tail;
  }
  return null;
}

/** ecitmatch first; fall back to a title-field esearch. */
export async function search(ref, email, opts = {}) {
  const pmids = [];
  const exact = await citationMatch(ref, email, opts);
  if (exact) pmids.push(exact);

  if (ref.title) {
    // [Title] restricts to the title field, which is far more precise than the
    // old code's untagged whole-reference term.
    let data = await get('pubmed', `${EUTILS}/esearch.fcgi`, {
      params: {
        db: 'pubmed', term: `"${ref.title}"[Title]`, retmax: 8, retmode: 'json', tool: TOOL, email,
      },
      ...opts,
    });
    let found = ((data || {}).esearchresult || {}).idlist || [];
    if (!found.length) {
      // loosen to all-fields when the exact title phrase isn't indexed
      data = await get('pubmed', `${EUTILS}/esearch.fcgi`, {
        params: {
          db: 'pubmed',
          term: `${ref.title} ${ref.lead_author || ''}`.trim(),
          retmax: 8,
          retmode: 'json',
          tool: TOOL,
          email,
        },
        ...opts,
      });
      found = ((data || {}).esearchresult || {}).idlist || [];
    }
    for (const p of found) if (!pmids.includes(p)) pmids.push(p);
  }

  const cands = await summaries(pmids.slice(0, 10), email, opts);
  for (const c of cands) {
    if (exact && c.pmid === exact) {
      // flag for the scorer: this one came from the deterministic matcher
      c.provider = 'pubmed:ecitmatch';
      c.providers = new Set(['pubmed:ecitmatch', 'pubmed']);
    }
  }
  return cands;
}

export async function byPmid(pmid, email, opts = {}) {
  const cands = await summaries([pmid], email, opts);
  return cands.length ? cands[0] : null;
}
